File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/python/YgriegaSB.py
"""
 * EJERCICIO:
 * - Crea ejemplos utilizando todos los tipos de operadores de tu lenguaje:
 *   Aritméticos, lógicos, de comparación, asignación, identidad, pertenencia, bits...
 *   (Ten en cuenta que cada lenguaje puede poseer unos diferentes)
 * - Utilizando las operaciones con operadores que tú quieras, crea ejemplos
 *   que representen todos los tipos de estructuras de control que existan
 *   en tu lenguaje:
 *   Condicionales, iterativas, excepciones...
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que imprima por consola todos los números comprendidos
 * entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
 *
 * Seguro que al revisar detenidamente las posibilidades has descubierto algo nuevo.
 """

# Suma
a = 5
b = 3
suma = a + b
print("Suma:", suma)  # Resultado: 8

# Resta
resta = a - b
print("Resta:", resta)  # Resultado: 2

# Multiplicación
multiplicacion = a * b
print("Multiplicación:", multiplicacion)  # Resultado: 15

# División
division = a / b
print("División:", division)  # Resultado: 1.6666666666666667

# División entera
division_entera = a // b
print("División entera:", division_entera)  # Resultado: 1

# Módulo (resto de la división)
modulo = a % b
print("Módulo:", modulo)  # Resultado: 2

# Exponenciación
exponenciacion = a ** b
print("Exponenciación:", exponenciacion)  # Resultado: 125

# AND lógico
x = True
y = False
and_logico = x and y
print("AND lógico:", and_logico)  # Resultado: False

# OR lógico
or_logico = x or y
print("OR lógico:", or_logico)  # Resultado: True

# NOT lógico
not_logico = not x
print("NOT lógico:", not_logico)  # Resultado: False

# Igualdad
a = 5
b = 3
igualdad = a == b
print("Igualdad:", igualdad)  # Resultado: False

# Desigualdad
desigualdad = a != b
print("Desigualdad:", desigualdad)  # Resultado: True

# Mayor que
mayor_que = a > b
print("Mayor que:", mayor_que)  # Resultado: True

# Menor que
menor_que = a < b
print("Menor que:", menor_que)  # Resultado: False

# Mayor o igual que
mayor_igual_que = a >= b
print("Mayor o igual que:", mayor_igual_que)  # Resultado: True

# Menor o igual que
menor_igual_que = a <= b
print("Menor o igual que:", menor_igual_que)  # Resultado: False

x = 5
# Asignación
x += 2  # Equivalente a: x = x + 2
print("Asignación:", x)  # Resultado: 7

x = [1, 2, 3]
y = [1, 2, 3]
z = x

# Identidad (comparando referencias de objetos)
identidad1 = x is y
print("Identidad 1:", identidad1)  # Resultado: False

identidad2 = x is z
print("Identidad 2:", identidad2)  # Resultado: True

# No Identidad
no_identidad = x is not y
print("No Identidad:", no_identidad)  # Resultado: True

# Pertenencia (comprobando si un valor está presente en una secuencia)
lista = [1, 2, 3, 4, 5]
pertenencia = 3 in lista
print("Pertenencia:", pertenencia)  # Resultado: True

# No Pertenencia
no_pertenencia = 6 not in lista
print("No Pertenencia:", no_pertenencia)  # Resultado: True

a = 10  # Representación binaria: 1010
b = 4   # Representación binaria: 0100

# AND a nivel de bits
and_bits = a & b
print("AND a nivel de bits:", and_bits)  # Resultado: 0 (binario: 0000)

# OR a nivel de bits
or_bits = a | b
print("OR a nivel de bits:", or_bits)    # Resultado: 14 (binario: 1110)

# XOR a nivel de bits
xor_bits = a ^ b
print("XOR a nivel de bits:", xor_bits)  # Resultado: 14 (binario: 1110)

# Desplazamiento a la izquierda
despl_izquierda = a << 2
print("Desplazamiento a la izquierda:", despl_izquierda)  # Resultado: 40 (binario: 101000)

# Desplazamiento a la derecha
despl_derecha = a >> 1
print("Desplazamiento a la derecha:", despl_derecha)  # Resultado: 5 (binario: 101)

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/YgriegaSB.js
/*
 * EJERCICIO:
 * - Crea ejemplos utilizando todos los tipos de operadores de tu lenguaje:
 *   Aritméticos, lógicos, de comparación, asignación, identidad, pertenencia, bits...
 *   (Ten en cuenta que cada lenguaje puede poseer unos diferentes)
 * - Utilizando las operaciones con operadores que tú quieras, crea ejemplos
 *   que representen todos los tipos de estructuras de control que existan
 *   en tu lenguaje:
 *   Condicionales, iterativas, excepciones...
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que imprima por consola todos los números comprendidos
 * entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
 *
 * Seguro que al revisar detenidamente las posibilidades has descubierto algo nuevo.
 */

// Aritméticos
let a = 1;
let b = 2;
let suma = a + b;
let resta = a - b;
let multiplicacion = a * b;
let division = a / b;
let modulo = a % b;
let incremento = a++;
let decremento = a--;

console.log('Suma = '+ suma);
console.log('Resta = '+ resta);
console.log('Multiplicación = '+ multiplicacion);
console.log('División = '+ division);
console.log('Módulo = '+ modulo);
console.log('Incremento = '+ incremento);
console.log('Decremento = '+ decremento);

// Asignacion

a = b;
console.log('a = b = '+ a);
a += 2;
console.log('Suma a = '+ a);
a -= 2;
console.log('Resta a = '+ a);
a *= 2;
console.log('Multiplicación a = '+ a);
a /= 2;
console.log('División a = '+ a);

// Comparación e identidad

console.log('a = '+a+' | b = '+ b);
let c = a > b;
console.log('a mayor que b -> '+c);
c = a < b;
console.log('a menor que b -> '+c);
c = a >= b;
console.log('a mayor o igual que b -> '+c);
c = a <= b;
console.log('a menor o igual que b -> '+c);
c = a == b;
console.log('a igual que b -> '+c);
c = a === b;
console.log('a estrictamente igual que b -> '+c);
c = a != b;
console.log('a diferente que b -> '+c);
c = a !== b;
console.log('a estrictamente diferente que b -> '+c);

// logicos

let combined = true && false;
console.log('AND -> '+ combined);
combined = true || false;
console.log('OR -> '+ combined);
combined = !true;
console.log('NOT -> '+ combined);

// Concatenar

let cadena1 = "Hola ";
let cadena2 = "mundo!";
let concatenada = cadena1 + cadena2;
console.log('Concatenación -> '+ concatenada);

// operadores ternarios

console.log(combined);
let resultado = combined===true? "Si" : "No";
console.log('Operador ternario -> '+ resultado);

// Tipos de datos
console.log(typeof(combined));
console.log(typeof(resultado));


// propagacion 
const numeros = [1, 2, 3, 4, 5];
console.log(...numeros); // Salida: 1 2 3 4 5

const numeros2 = [6, 7, 8];
const numerosCombinados = [...numeros, ...numeros2];
console.log(numerosCombinados); // Salida: [1, 2, 3, 4, 5, 6, 7, 8]


// membresia

    //in
    const persona = { nombre: 'Juan', edad: 30 };

    console.log('nombre' in persona); // true
    console.log('apellido' in persona); // false

    //instanceof
    class Persona {
        constructor(nombre) {
            this.nombre = nombre;
        }
    }
    
    const juan = new Persona('Juan');

    console.log(juan instanceof Persona); // true
    console.log(juan instanceof Array); // false
    
// exponenciacion
let potencia = a ** 4;
console.log(potencia);

// bits 
let x = 5; // Representación binaria: 0101

// Operador &= (AND a nivel de bits y asignación)
x &= 3; // Equivalente a: x = x & 3
// 0101 & 0011 = 0001 (resultado: 1)
console.log(x); // Salida: 1

// Operador |= (OR a nivel de bits y asignación)
x |= 8; // Equivalente a: x = x | 8
// 0001 | 1000 = 1001 (resultado: 9)
console.log(x); // Salida: 9

// Operador ^= (XOR a nivel de bits y asignación)
x ^= 12; // Equivalente a: x = x ^ 12
// 1001 ^ 1100 = 0101 (resultado: 5)
console.log(x); // Salida: 5

// Operador <<= (desplazamiento a la izquierda y asignación)
x <<= 2; // Equivalente a: x = x << 2
// 0101 << 2 = 010100 (resultado: 20)
console.log(x); // Salida: 20

// Operador >>= (desplazamiento a la derecha y asignación)
x >>= 1; // Equivalente a: x = x >> 1
// 010100 >> 1 = 001010 (resultado: 10)
console.log(x); // Salida: 10

// Operador >>>= (desplazamiento a la derecha sin signo y asignación)
x >>>= 1; // Equivalente a: x = x >>> 1
// 001010 >>> 1 = 000101 (resultado: 5)
console.log(x); // Salida: 5

// Estructuras de control

//if
let z = 26, y = 28;
if(z < y) {
    console.log('z es menor que y');
} else {
    console.log('No es menor que el');
}

// for

const animals = ['dog', 'cat', 'bird', 'horse']
for (let mascotas of animals) {
    //console.log(mascotas);
}

for (let i = 0; i <= animals.length; i++) {
    //console.log(animals[i]);
}

for (let name in animals) {
    //console.log('Indice '+name);
}

// while
let i = 0;
while(i < animals.length) {
    //console.log(animals[i]);
    i++;
}

// do while
let q = 0;
const nums = [29.8748, 30.1234, 30.1234, 3, 4, 5, 6, 7, 8, 9, 10]
do {
    //console.log(nums[q]);
    q++;
} while (q < nums.length);

// switch
let verdader = true;
let falso = false;
let input = undefined;

switch (input) {
    case verdader:
        console.log('Verdadero');
        break;
    case falso:
        console.log('Falso');
        break;
    default:
        console.log('No aplica');
        break;
}

// try/catch

try {
    let result = 10 / 0;
    console.log(result);
} catch (error) {
    console.log(error.message); // Mostrará "Division by zero"
} finally {
    console.log('Final');
}

// DIFICULTAD EXTRA (opcional):
// Crea un programa que imprima por consola todos los números comprendidos
// entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
console.log('\nSTART PROGRAM\n');
const output = [];
const programa = (i) => (i !== 16 && i % 3 !== 0 && i % 2 === 0) ? output.push(i) : null;

for (let i = 10; i <= 55; i++) programa(i);

console.log(output);
console.log('\nEND PROGRAM\n');
